### PR TITLE
Add support for generic properties in the Properties View

### LINKED ISSUE #53: Add support for generic options
Offer support for schema providers and schema decorators for the options displayed in the Properties view.
We should support passing the decorators and providers as a prop to the jsonforms-editor .

### PR DESCRIPTION
**Pull request by AlexandraBuzila**
- add support for schema providers and decorators for the PropertiesView
- schema providers and decorators can be passed as a prop to the editor

Fixes #53

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -46,14 +59,31 @@ const useStyles = makeStyles((theme) => ({
...
+];
+const defaultSchemaProviders: PropertiesSchemasProvider[] = [
+  propertiesSchemaProvider,
+];
+interface AppProps {
+  schemaProviders?: PropertiesSchemasProvider[];
+  schemaDecorators?: PropertiesSchemasDecorator[];
+}
+const App: React.FC<AppProps> = ({ schemaProviders, schemaDecorators }) => {
+  const decorators = schemaDecorators ?? defaultSchemaDecorators;
```
Comment (eneufeld) [inline src/App.tsx:76]: I'm not sure it is the best thing to define and use `defaultSchemaDecorators`, maybe we should pass the props rather using the index.tsx? @sdirix  what is your take on this?

File: src/properties/components/Properties.tsx [hunk 2]
```diff
@@ -74,6 +57,22 @@ export const Properties = () => {
     },
     [data, dispatch, uiElement]
   );
+  const propertiesService = usePropertiesService();
+
+  const getProperties = (
```
Comment (eneufeld) [inline src/properties/components/Properties.tsx:62]: shouldn't we wrap this in a `useEffect` hook?

File: src/properties/propertiesService.ts [hunk 3]
```diff
@@ -23,132 +18,71 @@ export interface PropertiesService {
...
-const createControl = (controlScope: string): ControlElement => ({
-  type: 'Control',
-  scope: controlScope,
-});
-
-interface PropertiesSchemasDecorators {
+/**
+ * Decorator for the PropertySchemas of an EditorUISchemaElement.
+ */
+export interface PropertiesSchemasDecorator {
```
Comment (eneufeld) [inline src/properties/propertiesService.ts:32]: do we expect this interface to provide more than this one function? if not maybe convert to a function interface?

File: src/properties/propertiesService.ts [hunk 4]
```diff
@@ -23,132 +18,71 @@ export interface PropertiesService {
...
+ * an EditorUISchemaElement.
+ */
+export const NOT_APPLICABLE = -1;
+
+/**
+ * Returns a PropertySchemas object for an EditorUISchemaElement. The tester will return a ranking
+ * or NOT_APPLICABLE if the provider cannot supply any schema for the given editor element. */
+export interface PropertiesSchemasProvider {
+  tester: (uiElement: EditorUISchemaElement) => number;
+  getPropertiesSchemas: (
```
Comment (eneufeld) [inline src/properties/propertiesService.ts:51]: shouldn't this be named: 
```suggestion
  getPropertySchemas: (
```

File: src/properties/propertiesService.ts [hunk 5]
```diff
@@ -23,132 +18,71 @@ export interface PropertiesService {
...
   getProperties = (
     uiElement: EditorUISchemaElement,
     schemaElement: SchemaElement | undefined
   ): PropertySchemas | undefined => {
-    const decoratedSchemas = decorators.reduce(
+    const providersWithRank = this.schemaProviders.map((provider) => ({
+      provider,
+      rank: provider.tester(uiElement),
+    }));
+    const highestRankedProvider = maxBy(providersWithRank, ({ rank }) => rank);
```
Comment (eneufeld) [inline src/properties/propertiesService.ts:69]: why not unify the map and maxBy?
```suggestion
    const provider = maxBy(this.schemaProviders, p => p.tester(uiElement));
```

File: src/properties/propertiesService.ts [hunk 6]
```diff
@@ -23,132 +18,71 @@ export interface PropertiesService {
...
+    const { provider, rank } = highestRankedProvider || {};
+    if (!provider || !rank || rank === NOT_APPLICABLE) {
```
Comment (eneufeld) [inline src/properties/propertiesService.ts:71]: of course you need to adapt this:
```suggestion
   if (provider === undefined || provider.tester(uiElement) === NOT_APPLICABLE) {
```

File: src/App.tsx (hunk 1 above)
Comment (AlexandraBuzila) [reply src/App.tsx]: Not sure if we should offer defaults or not. How will the other services be configured?

Comment (sdirix) [reply src/App.tsx]: When we expect all instances to set `schemaDecorators` anyway then I would go with Eugen's suggestion. If we expect that they will not be customized that often, then I'm completely fine with the optional approach. However I would use regular Javascript defaults, i.e.
```typescript
interface AppProps {
  schemaProviders: PropertySchemasProvider[];
  schemaDecorators?: PropertySchemasDecorator[];
}
const App: React.FC<AppProps> = ({
  schemaProviders,
  schemaDecorators = defaultSchemaDecorators,
}) => {
```
In the end I think both approaches are fine here.

**Pull request merged**